// @ionic/cloud/dist/es5/index.js
"use strict";
var auth_1 = require("./auth");
exports.Auth = auth_1.Auth;
exports.AuthType = auth_1.AuthType;
exports.BasicAuthType = auth_1.BasicAuthType;
exports.CustomAuthType = auth_1.CustomAuthType;
exports.FacebookAuth = auth_1.FacebookAuth;
exports.FacebookAuthType = auth_1.FacebookAuthType;
exports.GithubAuthType = auth_1.GithubAuthType;
exports.GoogleAuth = auth_1.GoogleAuth;
exports.GoogleAuthType = auth_1.GoogleAuthType;
exports.InstagramAuthType = auth_1.InstagramAuthType;
exports.LinkedInAuthType = auth_1.LinkedInAuthType;
exports.TwitterAuthType = auth_1.TwitterAuthType;
var client_1 = require("./client");
exports.Client = client_1.Client;
var config_1 = require("./config");
exports.Config = config_1.Config;
var cordova_1 = require("./cordova");
exports.Cordova = cordova_1.Cordova;
var core_1 = require("./core");
exports.Core = core_1.Core;
var deploy_1 = require("./deploy/deploy");
exports.Deploy = deploy_1.Deploy;
var device_1 = require("./device");
exports.Device = device_1.Device;
var database_1 = require("./database");
exports.Database = database_1.Database;
var errors_1 = require("./errors");
exports.Exception = errors_1.Exception;
exports.DetailedError = errors_1.DetailedError;
var di_1 = require("./di");
exports.DIContainer = di_1.Container;
var events_1 = require("./events");
exports.EventEmitter = events_1.EventEmitter;
var insights_1 = require("./insights");
exports.Insights = insights_1.Insights;
var logger_1 = require("./logger");
exports.Logger = logger_1.Logger;
var push_1 = require("./push/push");
exports.Push = push_1.Push;
var message_1 = require("./push/message");
exports.PushMessage = message_1.PushMessage;
var storage_1 = require("./storage");
exports.Storage = storage_1.Storage;
exports.LocalStorageStrategy = storage_1.LocalStorageStrategy;
exports.SessionStorageStrategy = storage_1.SessionStorageStrategy;
var user_1 = require("./user/user");
exports.UserContext = user_1.UserContext;
exports.User = user_1.User;
exports.SingleUserService = user_1.SingleUserService;
var angular_1 = require("./angular");
angular_1.bootstrapAngular1();
