// @ionic/db/lib/ast.js
'use strict';

exports.__esModule = true;
exports.UserDataTerm = exports.Limit = exports.Order = exports.Below = exports.Above = exports.FindAll = exports.Find = exports.Collection = exports.TermBase = undefined;

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

exports.applyChange = applyChange;

var _Observable = require('rxjs/Observable');

require('rxjs/add/observable/empty');

require('rxjs/add/operator/publishReplay');

require('rxjs/add/operator/scan');

require('rxjs/add/operator/filter');

require('rxjs/add/operator/map');

require('rxjs/add/operator/toArray');

require('rxjs/add/operator/defaultIfEmpty');

require('rxjs/add/operator/ignoreElements');

require('rxjs/add/operator/merge');

require('rxjs/add/operator/mergeMap');

require('rxjs/add/operator/take');

require('rxjs/add/operator/do');

var _snakeCase = require('snake-case');

var _snakeCase2 = _interopRequireDefault(_snakeCase);

var _deepEqual = require('deep-equal');

var _deepEqual2 = _interopRequireDefault(_deepEqual);

var _checkArgs = require('./util/check-args');

var _checkArgs2 = _interopRequireDefault(_checkArgs);

var _validIndexValue = require('./util/valid-index-value.js');

var _validIndexValue2 = _interopRequireDefault(_validIndexValue);

var _serialization = require('./serialization.js');

var _watchRewrites = require('./hacks/watch-rewrites');

var _watchRewrites2 = _interopRequireDefault(_watchRewrites);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var window = void 0;

function deepCompare(val1, val2) {
  if (window && window.angular) {
    return (0, _deepEqual2.default)(angular.fromJson(angular.toJson(val1)), angular.fromJson(angular.toJson(val1)));
  }
  return (0, _deepEqual2.default)(val1, val2);
}

/**
 @this TermBase

 Validation check to throw an exception if a method is chained onto a
 query that already has it. It belongs to TermBase, but we don't want
 to pollute the objects with it (since it isn't useful to api users),
 so it's dynamically bound with .call inside methods that use it.
*/
function checkIfLegalToChain(key) {
  if (this._legalMethods.indexOf(key) === -1) {
    throw new Error(key + ' cannot be called on the current query');
  }
  if ((0, _snakeCase2.default)(key) in this._query) {
    throw new Error(key + ' has already been called on this query');
  }
}

// Abstract base class for terms

var TermBase = exports.TermBase = function () {
  function TermBase(sendRequest, query, legalMethods, digest) {
    (0, _classCallCheck3.default)(this, TermBase);

    this._sendRequest = sendRequest;
    this._query = query;
    this._legalMethods = legalMethods;
    this._digest = digest;
    if (!this._digest) {
      this._digest = function () {};
    }
  }

  TermBase.prototype.toString = function toString() {
    var string = 'Collection(\'' + this._query.collection + '\')';
    if (this._query.find) {
      string += '.find(' + JSON.stringify(this._query.find) + ')';
    }
    if (this._query.find_all) {
      string += '.findAll(' + JSON.stringify(this._query.find_all) + ')';
    }
    if (this._query.order) {
      string += '.order(' + JSON.stringify(this._query.order[0]) + ', ' + (JSON.stringify(this._query.order[1]) + ')');
    }
    if (this._query.above) {
      string += '.above(' + JSON.stringify(this.query.above[0]) + ', ' + (JSON.stringify(this.query.above[1]) + ')');
    }
    if (this._query.below) {
      string += '.below(' + JSON.stringify(this.query.below[0]) + ', ' + (JSON.stringify(this.query.below[1]) + ')');
    }
    if (this._query.limit) {
      string += '.limit(this._query.limit))';
    }
    return string;
  };
  // Returns a sequence of the result set. Every time it changes the
  // updated sequence will be emitted. If raw change objects are
  // needed, pass the option 'rawChanges: true'. An observable is
  // returned which will lazily emit the query when it is subscribed
  // to


  TermBase.prototype.watch = function watch() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$rawChanges = _ref.rawChanges,
        rawChanges = _ref$rawChanges === undefined ? false : _ref$rawChanges;

    var query = (0, _watchRewrites2.default)(this, this._query);
    var raw = this._sendRequest('subscribe', query);
    if (rawChanges) {
      return raw.do(this._digest);
    } else {
      return makePresentable(raw, this._query).share().do(this._digest);
    }
  };
  // Grab a snapshot of the current query (non-changefeed). Emits an
  // array with all results. An observable is returned which will
  // lazily emit the query when subscribed to


  TermBase.prototype.fetch = function fetch() {
    var raw = this._sendRequest('query', this._query).map(function (val) {
      delete val.$hz_v$;
      return val;
    });
    if (this._query.find) {
      return raw.defaultIfEmpty(null).do(this._digest);
    } else {
      return raw.toArray().do(this._digest);
    }
  };

  TermBase.prototype.findAll = function findAll() {
    for (var _len = arguments.length, fieldValues = Array(_len), _key = 0; _key < _len; _key++) {
      fieldValues[_key] = arguments[_key];
    }

    checkIfLegalToChain.call(this, 'findAll');
    (0, _checkArgs2.default)('findAll', arguments, { maxArgs: 100 });
    return new FindAll(this._sendRequest, this._query, fieldValues, this._digest);
  };

  TermBase.prototype.find = function find(idOrObject) {
    checkIfLegalToChain.call(this, 'find');
    (0, _checkArgs2.default)('find', arguments);
    return new Find(this._sendRequest, this._query, idOrObject, this._digest);
  };

  TermBase.prototype.order = function order(fields) {
    var direction = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'ascending';

    checkIfLegalToChain.call(this, 'order');
    (0, _checkArgs2.default)('order', arguments, { minArgs: 1, maxArgs: 2 });
    return new Order(this._sendRequest, this._query, fields, direction, this._digest);
  };

  TermBase.prototype.above = function above(aboveSpec) {
    var bound = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'closed';

    checkIfLegalToChain.call(this, 'above');
    (0, _checkArgs2.default)('above', arguments, { minArgs: 1, maxArgs: 2 });
    return new Above(this._sendRequest, this._query, aboveSpec, bound, this._digest);
  };

  TermBase.prototype.below = function below(belowSpec) {
    var bound = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'open';

    checkIfLegalToChain.call(this, 'below');
    (0, _checkArgs2.default)('below', arguments, { minArgs: 1, maxArgs: 2 });
    return new Below(this._sendRequest, this._query, belowSpec, bound, this._digest);
  };

  TermBase.prototype.limit = function limit(size) {
    checkIfLegalToChain.call(this, 'limit');
    (0, _checkArgs2.default)('limit', arguments);
    return new Limit(this._sendRequest, this._query, size, this._digest);
  };

  return TermBase;
}();

// Turn a raw observable of server responses into user-presentable events
//
// `observable` is the base observable with full responses coming from
//              the IonicDBSocket
// `query` is the value of `options` in the request


function makePresentable(observable, query) {
  // Whether the entire data structure is in each change
  var pointQuery = Boolean(query.find);

  if (pointQuery) {
    var _ret = function () {
      var hasEmitted = false;
      var seedVal = { val: null, prev: {} };
      // Simplest case: just pass through new_val
      return {
        v: observable.filter(function (change) {
          return !hasEmitted || change.type !== 'state';
        }).scan(function (state, change) {
          hasEmitted = true;
          if (change.new_val != null) {
            delete change.new_val.$hz_v$;
          }
          if (change.old_val != null) {
            delete change.old_val.$hz_v$;
          }
          if (change.state === 'synced') {
            return state;
          } else {
            state.prev = state.val;
            state.val = change.new_val;
            return state;
          }
        }, seedVal).filter(function (state) {
          return !deepCompare(state.val, state.prev);
        }).map(function (x) {
          return x.val;
        })
      };
    }();

    if ((typeof _ret === 'undefined' ? 'undefined' : (0, _typeof3.default)(_ret)) === "object") return _ret.v;
  } else {
    var _seedVal = { emitted: false, val: [], prev: null };
    return observable.scan(function (state, change) {
      if (change.new_val != null) {
        delete change.new_val.$hz_v$;
      }
      if (change.old_val != null) {
        delete change.old_val.$hz_v$;
      }
      if (change.state === 'synced') {
        state.emitted = true;
      }
      if (change.state === 'reconnecting') {
        state.emitted = false;
        state.prev = state.val;
        state.val = [];
      }
      state.val = applyChange(state.val.slice(), change);
      return state;
    }, _seedVal).filter(function (state) {
      return state.emitted && !deepCompare(state.val, state.prev);
    }).map(function (x) {
      return x.val;
    });
  }
}

function applyChange(arr, change) {
  switch (change.type) {
    case 'remove':
    case 'uninitial':
      {
        // Remove old values from the array
        if (change.old_offset != null) {
          arr.splice(change.old_offset, 1);
        } else {
          var index = arr.findIndex(function (x) {
            return deepCompare(x.id, change.old_val.id);
          });
          if (index === -1) {
            // Programming error. This should not happen
            throw new Error('change couldn\'t be applied: ' + JSON.stringify(change));
          }
          arr.splice(index, 1);
        }
        break;
      }
    case 'add':
    case 'initial':
      {
        // Add new values to the array
        if (change.new_offset != null) {
          // If we have an offset, put it in the correct location
          arr.splice(change.new_offset, 0, change.new_val);
        } else {
          // otherwise for unordered results, push it on the end
          arr.push(change.new_val);
        }
        break;
      }
    case 'change':
      {
        // Modify in place if a change is happening
        if (change.old_offset != null) {
          // Remove the old document from the results
          arr.splice(change.old_offset, 1);
        }
        if (change.new_offset != null) {
          // Splice in the new val if we have an offset
          arr.splice(change.new_offset, 0, change.new_val);
        } else {
          // If we don't have an offset, find the old val and
          // replace it with the new val
          var _index = arr.findIndex(function (x) {
            return deepCompare(x.id, change.old_val.id);
          });
          if (_index === -1) {
            // indicates a programming bug. The server gives us the
            // ordering, so if we don't find the id it means something is
            // buggy.
            throw new Error('change couldn\'t be applied: ' + JSON.stringify(change));
          }
          arr[_index] = change.new_val;
        }
        break;
      }
    case 'state':
      {
        // This gets hit if we have not emitted yet, and should
        // result in an empty array being output.
        break;
      }
    default:
      throw new Error('unrecognized \'type\' field from server ' + JSON.stringify(change));
  }
  return arr;
}

/** @this Collection
 Implements writeOps for the Collection class
*/
function writeOp(name, args, documents, digest) {
  (0, _checkArgs2.default)(name, args);
  var isBatch = true;
  var wrappedDocs = documents;
  if (!Array.isArray(documents)) {
    // Wrap in an array if we need to
    wrappedDocs = [documents];
    isBatch = false;
  } else if (documents.length === 0) {
    // Don't bother sending no-ops to the server
    return _Observable.Observable.empty();
  }
  var options = Object.assign({}, this._query, { data: (0, _serialization.serialize)(wrappedDocs) });
  var observable = this._sendRequest(name, options);
  if (isBatch) {
    // If this is a batch writeOp, each document may succeed or fail
    // individually.
    observable = observable.map(function (resp) {
      return resp.error ? new Error(resp.error) : resp;
    });
  } else {
    (function () {
      // If this is a single writeOp, the entire operation should fail
      // if any fails.
      var _prevOb = observable;
      observable = _Observable.Observable.create(function (subscriber) {
        _prevOb.subscribe({
          next: function next(resp) {
            if (resp.error) {
              // TODO: handle error ids when we get them
              subscriber.error(new Error(resp.error));
            } else {
              subscriber.next(resp);
            }
          },
          error: function error(err) {
            subscriber.error(err);
          },
          complete: function complete() {
            subscriber.complete();
          }
        });
      });
    })();
  }
  if (!this._lazyWrites) {
    // Need to buffer response since this becomes a hot observable and
    // when we subscribe matters
    observable = observable.publishReplay().refCount();
    observable.subscribe();
  }
  return observable.do(this._digest);
}

var Collection = exports.Collection = function (_TermBase) {
  (0, _inherits3.default)(Collection, _TermBase);

  function Collection(sendRequest, collectionName, lazyWrites, digest) {
    (0, _classCallCheck3.default)(this, Collection);

    var query = { collection: collectionName };
    var legalMethods = ['find', 'findAll', 'order', 'above', 'below', 'limit'];

    var _this = (0, _possibleConstructorReturn3.default)(this, _TermBase.call(this, sendRequest, query, legalMethods, digest));

    _this._lazyWrites = lazyWrites;
    return _this;
  }

  Collection.prototype.store = function store(documents) {
    return writeOp.call(this, 'store', arguments, documents);
  };

  Collection.prototype.upsert = function upsert(documents) {
    return writeOp.call(this, 'upsert', arguments, documents);
  };

  Collection.prototype.insert = function insert(documents) {
    return writeOp.call(this, 'insert', arguments, documents);
  };

  Collection.prototype.replace = function replace(documents) {
    return writeOp.call(this, 'replace', arguments, documents);
  };

  Collection.prototype.update = function update(documents) {
    return writeOp.call(this, 'update', arguments, documents);
  };

  Collection.prototype.remove = function remove(documentOrId) {
    var wrapped = (0, _validIndexValue2.default)(documentOrId) ? { id: documentOrId } : documentOrId;
    return writeOp.call(this, 'remove', arguments, wrapped);
  };

  Collection.prototype.removeAll = function removeAll(documentsOrIds) {
    if (!Array.isArray(documentsOrIds)) {
      throw new Error('removeAll takes an array as an argument');
    }
    var wrapped = documentsOrIds.map(function (item) {
      if ((0, _validIndexValue2.default)(item)) {
        return { id: item };
      } else {
        return item;
      }
    });
    return writeOp.call(this, 'removeAll', arguments, wrapped);
  };

  return Collection;
}(TermBase);

var Find = exports.Find = function (_TermBase2) {
  (0, _inherits3.default)(Find, _TermBase2);

  function Find(sendRequest, previousQuery, idOrObject, digest) {
    (0, _classCallCheck3.default)(this, Find);

    var findObject = (0, _validIndexValue2.default)(idOrObject) ? { id: idOrObject } : idOrObject;
    var query = Object.assign({}, previousQuery, { find: findObject });
    return (0, _possibleConstructorReturn3.default)(this, _TermBase2.call(this, sendRequest, query, [], digest));
  }

  return Find;
}(TermBase);

var FindAll = exports.FindAll = function (_TermBase3) {
  (0, _inherits3.default)(FindAll, _TermBase3);

  function FindAll(sendRequest, previousQuery, fieldValues, digest) {
    (0, _classCallCheck3.default)(this, FindAll);

    var wrappedFields = fieldValues.map(function (item) {
      return (0, _validIndexValue2.default)(item) ? { id: item } : item;
    });
    var options = { find_all: wrappedFields };
    var findAllQuery = Object.assign({}, previousQuery, options);
    var legalMethods = void 0;
    if (wrappedFields.length === 1) {
      legalMethods = ['order', 'above', 'below', 'limit'];
    } else {
      // The vararg version of findAll cannot have anything chained to it
      legalMethods = [];
    }
    return (0, _possibleConstructorReturn3.default)(this, _TermBase3.call(this, sendRequest, findAllQuery, legalMethods, digest));
  }

  return FindAll;
}(TermBase);

var Above = exports.Above = function (_TermBase4) {
  (0, _inherits3.default)(Above, _TermBase4);

  function Above(sendRequest, previousQuery, aboveSpec, bound, digest) {
    (0, _classCallCheck3.default)(this, Above);

    var option = { above: [aboveSpec, bound] };
    var query = Object.assign({}, previousQuery, option);
    var legalMethods = ['findAll', 'order', 'below', 'limit'];
    return (0, _possibleConstructorReturn3.default)(this, _TermBase4.call(this, sendRequest, query, legalMethods, digest));
  }

  return Above;
}(TermBase);

var Below = exports.Below = function (_TermBase5) {
  (0, _inherits3.default)(Below, _TermBase5);

  function Below(sendRequest, previousQuery, belowSpec, bound, digest) {
    (0, _classCallCheck3.default)(this, Below);

    var options = { below: [belowSpec, bound] };
    var query = Object.assign({}, previousQuery, options);
    var legalMethods = ['findAll', 'order', 'above', 'limit'];
    return (0, _possibleConstructorReturn3.default)(this, _TermBase5.call(this, sendRequest, query, legalMethods, digest));
  }

  return Below;
}(TermBase);

var Order = exports.Order = function (_TermBase6) {
  (0, _inherits3.default)(Order, _TermBase6);

  function Order(sendRequest, previousQuery, fields, direction, digest) {
    (0, _classCallCheck3.default)(this, Order);

    var wrappedFields = Array.isArray(fields) ? fields : [fields];
    var options = { order: [wrappedFields, direction] };
    var query = Object.assign({}, previousQuery, options);
    var legalMethods = ['findAll', 'above', 'below', 'limit'];
    return (0, _possibleConstructorReturn3.default)(this, _TermBase6.call(this, sendRequest, query, legalMethods, digest));
  }

  return Order;
}(TermBase);

var Limit = exports.Limit = function (_TermBase7) {
  (0, _inherits3.default)(Limit, _TermBase7);

  function Limit(sendRequest, previousQuery, size, digest) {
    (0, _classCallCheck3.default)(this, Limit);

    var query = Object.assign({}, previousQuery, { limit: size });
    // Nothing is legal to chain after .limit
    return (0, _possibleConstructorReturn3.default)(this, _TermBase7.call(this, sendRequest, query, [], digest));
  }

  return Limit;
}(TermBase);

var UserDataTerm = exports.UserDataTerm = function () {
  function UserDataTerm(hz, handshake, socket) {
    (0, _classCallCheck3.default)(this, UserDataTerm);

    this._hz = hz;
    this._before = socket.ignoreElements().merge(handshake);
  }

  UserDataTerm.prototype._query = function _query(userId) {
    return this._hz.collection('users').find(userId);
  };

  UserDataTerm.prototype.fetch = function fetch() {
    var _this8 = this;

    return this._before.mergeMap(function (handshake) {
      if (handshake.id == null) {
        throw new Error('Unauthenticated users have no user document');
      } else {
        return _this8._query(handshake.id).fetch();
      }
    }).take(1); // necessary so that we complete, since _before is
    // infinite
  };

  UserDataTerm.prototype.watch = function watch() {
    var _this9 = this;

    for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    return this._before.mergeMap(function (handshake) {
      if (handshake.id === null) {
        throw new Error('Unauthenticated users have no user document');
      } else {
        var _query2;

        return (_query2 = _this9._query(handshake.id)).watch.apply(_query2, args);
      }
    });
  };

  return UserDataTerm;
}();
