// @ionic/db/lib/auth.js
'use strict';

exports.__esModule = true;
exports.TokenStorage = exports.FakeStorage = undefined;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

exports.authEndpoint = authEndpoint;
exports.clearAuthTokens = clearAuthTokens;

var _queryParse = require('./util/query-parse');

var _queryParse2 = _interopRequireDefault(_queryParse);

var _Observable = require('rxjs/Observable');

require('rxjs/add/operator/do');

require('rxjs/add/operator/map');

require('rxjs/add/observable/dom/ajax');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var HORIZON_JWT = 'ionicdb-jwt';

/** @this Horizon **/
function authEndpoint(name) {
  var _this = this;

  var endpointForName = function endpointForName(methods) {
    if (methods.hasOwnProperty(name)) {
      return _this._root + methods[name];
    } else {
      throw new Error('Unconfigured auth type: ' + name);
    }
  };
  if (!this._authMethods) {
    return _Observable.Observable.ajax(this._horizonPath + '/auth_methods').map(function (ajax) {
      return ajax.response;
    }).do(function (authMethods) {
      _this._authMethods = authMethods;
    }).map(endpointForName);
  } else {
    return _Observable.Observable.of(this._authMethods).map(endpointForName);
  }
}

// Simple shim to make a Map look like local/session storage

var FakeStorage = exports.FakeStorage = function () {
  function FakeStorage() {
    (0, _classCallCheck3.default)(this, FakeStorage);
    this._storage = new Map();
  }

  FakeStorage.prototype.setItem = function setItem(a, b) {
    return this._storage.set(a, b);
  };

  FakeStorage.prototype.getItem = function getItem(a) {
    return this._storage.get(a);
  };

  FakeStorage.prototype.removeItem = function removeItem(a) {
    return this._storage.delete(a);
  };

  return FakeStorage;
}();

function getStorage() {
  var storeLocally = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;

  var storage = void 0;
  try {
    if (!storeLocally || (typeof window === 'undefined' ? 'undefined' : (0, _typeof3.default)(window)) !== 'object' || window.localStorage === undefined) {
      storage = new FakeStorage();
    } else {
      // Mobile safari in private browsing has a localStorage, but it
      // has a size limit of 0
      window.localStorage.setItem('$$fake', 1);
      window.localStorage.removeItem('$$fake');
      storage = window.localStorage;
    }
  } catch (error) {
    if (window.sessionStorage === undefined) {
      storage = new FakeStorage();
    } else {
      storage = window.sessionStorage;
    }
  }
  return storage;
}

var TokenStorage = exports.TokenStorage = function () {
  function TokenStorage() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$authType = _ref.authType,
        authType = _ref$authType === undefined ? 'token' : _ref$authType,
        _ref$storage = _ref.storage,
        storage = _ref$storage === undefined ? getStorage(authType.storeLocally) : _ref$storage,
        _ref$path = _ref.path,
        path = _ref$path === undefined ? 'horizon' : _ref$path;

    (0, _classCallCheck3.default)(this, TokenStorage);

    this._storage = storage;
    this._path = path;
    if (typeof authType === 'string') {
      this._authType = authType;
    } else {
      this._authType = 'token';
      this.set(authType.token);
    }
  }

  TokenStorage.prototype._getHash = function _getHash() {
    var val = this._storage.getItem(HORIZON_JWT);
    if (val == null) {
      return {};
    } else {
      return JSON.parse(val);
    }
  };

  TokenStorage.prototype._setHash = function _setHash(hash) {
    this._storage.setItem(HORIZON_JWT, JSON.stringify(hash));
  };

  TokenStorage.prototype.set = function set(jwt) {
    var current = this._getHash();
    current[this._path] = jwt;
    this._setHash(current);
  };

  TokenStorage.prototype.get = function get() {
    return this._getHash()[this._path];
  };

  TokenStorage.prototype.remove = function remove() {
    var current = this._getHash();
    delete current[this._path];
    this._setHash(current);
  };

  TokenStorage.prototype.setAuthFromQueryParams = function setAuthFromQueryParams() {
    var parsed = typeof window !== 'undefined' && typeof window.location !== 'undefined' ? (0, _queryParse2.default)(window.location.search) : {};

    if (parsed.horizon_token != null) {
      this.set(parsed.horizon_token);
    }
  };

  // Handshake types are implemented here


  TokenStorage.prototype.handshake = function handshake() {
    // If we have a token, we should send it rather than requesting a
    // new one
    var token = this.get();
    if (token != null && this._authType === 'token') {
      return { method: 'token', token: token };
    } else if (this._authType === 'token') {
      throw new Error('Attempting authenticated login but no token detected. Make sure you\'ve logged in before attempting to call connect or make a query.');
    } else {
      return { method: this._authType };
    }
  };

  // Whether there is an auth token for the provided authType


  TokenStorage.prototype.hasAuthToken = function hasAuthToken() {
    var token = this.get();
    if (!token) {
      return false;
    }
    try {
      var meta = JSON.parse(atob(token.split('.')[1]));
      var exp = meta.exp;
      var now = new Date().getTime() / 1000;
      return now < exp;
    } catch (e) {
      return false;
    }
  };

  return TokenStorage;
}();

function clearAuthTokens() {
  return getStorage().removeItem(HORIZON_JWT);
}
