// @ionic/db/lib/model.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

exports.aggregate = aggregate;
exports.model = model;

var _Observable = require('rxjs/Observable');

require('rxjs/add/observable/of');

require('rxjs/add/observable/forkJoin');

require('rxjs/add/observable/combineLatest');

require('rxjs/add/operator/map');

var _isPlainObject = require('is-plain-object');

var _isPlainObject2 = _interopRequireDefault(_isPlainObject);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Unlike normal queries' .watch(), we don't support rawChanges: true
// for aggregates
function checkWatchArgs(args) {
  if (args.length > 0) {
    throw new Error(".watch() on aggregates doesn't support arguments!");
  }
}

// Other imports


function isTerm(term) {
  return typeof term.fetch === 'function' && typeof term.watch === 'function';
}

function isPromise(term) {
  return typeof term.then === 'function';
}

function isObservable(term) {
  return typeof term.subscribe === 'function' && typeof term.lift === 'function';
}

// Whether an object is primitive. We consider functions
// non-primitives, lump Dates and ArrayBuffers into primitives.
function isPrimitive(value) {
  if (value === null) {
    return true;
  }
  if (value === undefined) {
    return false;
  }
  if (typeof value === 'function') {
    return false;
  }
  if (['boolean', 'number', 'string'].indexOf(typeof value === 'undefined' ? 'undefined' : (0, _typeof3.default)(value)) !== -1) {
    return true;
  }
  if (value instanceof Date || value instanceof ArrayBuffer) {
    return true;
  }
  return false;
}

// Simple wrapper for primitives. Just emits the primitive

var PrimitiveTerm = function () {
  function PrimitiveTerm(value) {
    (0, _classCallCheck3.default)(this, PrimitiveTerm);

    this._value = value;
  }

  PrimitiveTerm.prototype.toString = function toString() {
    return this._value.toString();
  };

  PrimitiveTerm.prototype.fetch = function fetch() {
    return _Observable.Observable.of(this._value);
  };

  PrimitiveTerm.prototype.watch = function watch() {
    for (var _len = arguments.length, watchArgs = Array(_len), _key = 0; _key < _len; _key++) {
      watchArgs[_key] = arguments[_key];
    }

    checkWatchArgs(watchArgs);
    return _Observable.Observable.of(this._value);
  };

  return PrimitiveTerm;
}();

// Simple wrapper for observables to normalize the
// interface. Everything in an aggregate tree should be one of these
// term-likes


var ObservableTerm = function () {
  function ObservableTerm(value) {
    (0, _classCallCheck3.default)(this, ObservableTerm);

    this._value = value;
  }

  ObservableTerm.prototype.toString = function toString() {
    return this._value.toString();
  };

  ObservableTerm.prototype.fetch = function fetch() {
    return _Observable.Observable.from(this._value);
  };

  ObservableTerm.prototype.watch = function watch() {
    for (var _len2 = arguments.length, watchArgs = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      watchArgs[_key2] = arguments[_key2];
    }

    checkWatchArgs(watchArgs);
    return _Observable.Observable.from(this._value);
  };

  return ObservableTerm;
}();

// Handles aggregate syntax like [ query1, query2 ]


var ArrayTerm = function () {
  function ArrayTerm(value) {
    (0, _classCallCheck3.default)(this, ArrayTerm);

    // Ensure this._value is an array of Term
    this._value = value.map(function (x) {
      return aggregate(x);
    });
  }

  ArrayTerm.prototype._reducer = function _reducer() {
    for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      args[_key3] = arguments[_key3];
    }

    return args;
  };

  ArrayTerm.prototype._query = function _query(operation) {
    return this._value.map(function (x) {
      return x[operation]();
    });
  };

  ArrayTerm.prototype.toString = function toString() {
    return '[ ' + this._query('toString').join(', ') + ' ]';
  };

  ArrayTerm.prototype.fetch = function fetch() {
    if (this._value.length === 0) {
      return _Observable.Observable.empty();
    }

    var qs = this._query('fetch');
    return _Observable.Observable.forkJoin.apply(_Observable.Observable, qs.concat([this._reducer]));
  };

  ArrayTerm.prototype.watch = function watch() {
    for (var _len4 = arguments.length, watchArgs = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      watchArgs[_key4] = arguments[_key4];
    }

    checkWatchArgs(watchArgs);

    if (this._value.length === 0) {
      return _Observable.Observable.empty();
    }

    var qs = this._query('watch');
    return _Observable.Observable.combineLatest.apply(_Observable.Observable, qs.concat([this._reducer]));
  };

  return ArrayTerm;
}();

var AggregateTerm = function () {
  function AggregateTerm(value) {
    (0, _classCallCheck3.default)(this, AggregateTerm);

    // Ensure this._value is an array of [ key, Term ] pairs
    this._value = Object.keys(value).map(function (k) {
      return [k, aggregate(value[k])];
    });
  }

  AggregateTerm.prototype._reducer = function _reducer() {
    for (var _len5 = arguments.length, pairs = Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
      pairs[_key5] = arguments[_key5];
    }

    return pairs.reduce(function (prev, _ref) {
      var k = _ref[0],
          x = _ref[1];

      prev[k] = x;
      return prev;
    }, {});
  };

  AggregateTerm.prototype._query = function _query(operation) {
    return this._value.map(function (_ref2) {
      var k = _ref2[0],
          term = _ref2[1];
      return term[operation]().map(function (x) {
        return [k, x];
      });
    });
  };

  AggregateTerm.prototype.toString = function toString() {
    var s = this._value.map(function (_ref3) {
      var k = _ref3[0],
          term = _ref3[1];
      return '\'' + k + '\': ' + term;
    });
    return '{ ' + s.join(', ') + ' }';
  };

  AggregateTerm.prototype.fetch = function fetch() {
    if (this._value.length === 0) {
      return _Observable.Observable.of({});
    }

    var qs = this._query('fetch');
    return _Observable.Observable.forkJoin.apply(_Observable.Observable, qs.concat([this._reducer]));
  };

  AggregateTerm.prototype.watch = function watch() {
    for (var _len6 = arguments.length, watchArgs = Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
      watchArgs[_key6] = arguments[_key6];
    }

    checkWatchArgs(watchArgs);

    if (this._value.length === 0) {
      return _Observable.Observable.of({});
    }

    var qs = this._query('watch');
    return _Observable.Observable.combineLatest.apply(_Observable.Observable, qs.concat([this._reducer]));
  };

  return AggregateTerm;
}();

function aggregate(spec) {
  if (isTerm(spec)) {
    return spec;
  }
  if (isObservable(spec) || isPromise(spec)) {
    return new ObservableTerm(spec);
  }
  if (isPrimitive(spec)) {
    return new PrimitiveTerm(spec);
  }
  if (Array.isArray(spec)) {
    return new ArrayTerm(spec);
  }
  if ((0, _isPlainObject2.default)(spec)) {
    return new AggregateTerm(spec);
  }

  throw new Error('Can\'t make an aggregate with ' + spec + ' in it');
}

function model(constructor) {
  return function () {
    return aggregate(constructor.apply(undefined, arguments));
  };
}
