// @ionic/app-scripts/dist/util/constants.js
"use strict";
exports.FILE_CHANGE_EVENT = 'change';
exports.FILE_ADD_EVENT = 'add';
exports.FILE_DELETE_EVENT = 'unlink';
exports.DIRECTORY_ADD_EVENT = 'addDir';
exports.DIRECTORY_DELETE_EVENT = 'unlinkDir';
exports.SOURCE_MAP_TYPE_CHEAP = 'eval';
exports.SOURCE_MAP_TYPE_EXPENSIVE = 'source-map';
exports.BUILD_DIR = 'build';
exports.SRC_DIR = 'src';
exports.WWW_DIR = 'www';
exports.WWW_INDEX_FILENAME = 'index.html';
exports.ENV_VAR_PROD = 'prod';
exports.ENV_VAR_DEV = 'dev';
exports.ENV_VAR_IONIC_ENV = 'IONIC_ENV';
exports.ENV_VAR_ROOT_DIR = 'IONIC_ROOT_DIR';
exports.ENV_VAR_SRC_DIR = 'IONIC_SRC_DIR';
exports.ENV_VAR_WWW_DIR = 'IONIC_WWW_DIR';
exports.ENV_VAR_BUILD_DIR = 'IONIC_BUILD_DIR';
exports.ENV_VAR_APP_SCRIPTS_DIR = 'IONIC_APP_SCRIPTS_DIR';
exports.ENV_VAR_GENERATE_SOURCE_MAP = 'IONIC_GENERATE_SOURCE_MAP';
exports.ENV_VAR_SOURCE_MAP_TYPE = 'IONIC_SOURCE_MAP_TYPE';
exports.ENV_TS_CONFIG = 'IONIC_TS_CONFIG';
exports.ENV_APP_ENTRY_POINT = 'IONIC_APP_ENTRY_POINT';
exports.ENV_GLOB_UTIL = 'IONIC_GLOB_UTIL';
exports.ENV_CLEAN_BEFORE_COPY = 'IONIC_CLEAN_BEFORE_COPY';
exports.ENV_CLOSURE_JAR = 'IONIC_CLOSURE_JAR';
exports.ENV_OUTPUT_JS_FILE_NAME = 'IONIC_OUTPUT_JS_FILE_NAME';
exports.ENV_OUTPUT_JS_MAP_FILE_NAME = 'IONIC_OUTPUT_JS_MAP_FILE_NAME';
exports.ENV_OUTPUT_CSS_FILE_NAME = 'IONIC_OUTPUT_CSS_FILE_NAME';
exports.ENV_OUTPUT_CSS_MAP_FILE_NAME = 'IONIC_OUTPUT_CSS_MAP_FILE_NAME';
exports.ENV_WEBPACK_FACTORY = 'IONIC_WEBPACK_FACTORY';
exports.ENV_WEBPACK_LOADER = 'IONIC_WEBPACK_LOADER';
exports.BUNDLER_ROLLUP = 'rollup';
exports.BUNDLER_WEBPACK = 'webpack';
